template.tpl

```
@@ -592,37 +601,46 @@ const stdLogInfo = {
   traceId: traceIdHeader,
   eventName: eventData.event_name,
 };
-const url = standardEndpoint + '/import/environments/' + data.clientSideId + '/metrics';
-const requestHeaders = {
-  headers: { 
-    'Content-Type': 'application/json', 
-    'X-LaunchDarkly-Event-Schema': 4, 
-    'LD-API-Version': 'beta', 
+const url =
+  standardEndpoint + '/import/environments/' + data.clientSideId + '/metrics';
+
+// Note on version by LaunchDarkly docs: can be any format,
+// but you should update it if you make major changes to your implementation.
+// For now we hardcode it to '1'.
+const version = '1';
+const requestOptions = {
+  headers: {
+    'Content-Type': 'application/json',
+    'X-LaunchDarkly-Event-Schema': 4,
+    'LD-API-Version': 'beta',
     'X-LaunchDarkly-Payload-ID': insertId,
-    'User-Agent': 'MetricImport-' + data.companyName + '-int/1'
+    'User-Agent': 'MetricImport-' + data.companyName + '-int/' + version,
   },
   method: 'POST',
   timeout: 5000,
 };
 
-let ldEvent = {
+const ldEvent = {
   kind: 'custom',
   key: data.eventName,
-  creationDate: getTimestamp(),
-  contextKeys: { user: data.userValueDropDown === 'userId' ? eventData.userId : eventData[data.userValueCustom] }
+  creationDate: getTimestamp(data),
+  contextKeys: {
+    user:
+      data.userValueDropDown === 'userId' ? eventData.user_id : getEventData(data.userValueCustom),
+  },
 };
 
 if (data.metricType === 'metric') {
-  ldEvent.metricValue = makeNumber(eventData[data.metricValueCustom]);
+  ldEvent.metricValue = makeNumber(getEventData(data.metricValueCustom));
```

Docs say `metricValue` is required for metrics but LD won't validate the payload, should _we_ be checking it's a number and erroring if not? Or does `makeNumber()` already fail like this?

---

`makeNumber` will not fail. What it returns is always of type `number` (per `getType`) but it can "mean" `NaN`.
It is indeed an option to fail here. Would we want this to be a concern of the tag?

---

`NaN` will `JSON.stringify()` to `null` though. I think maybe we should fail here since the only alternative is silently failing on the LD side; better for someone to see the errors and have a chance to specify a default in their configuration or something, I think.

---

Last comment, promise. :)

I feel like we should still be checking the result of `makeNumber` rather than the input; if it's a string or something it should work unless it's nonsense, right?

Maybe like:
```javascript
  const metricValue = makeNumber(getEventData(data.metricValueCustom));

  if (JSON.stringify(metricValue) === "null") { // check for undefined, NaN, Infinity, etc
    return fail(loggingEnabled, stdLogInfo, {
      msg: 'Metric Value must be a number.',
    });
  }
  ldEvent.metricValue = metricValue;
```

---

You are right! If user provides a string that can be parsed into a number, then yes, indeed, we should be able to use it.
There is a lot of pain with `makeNumber`, especially if there are no guarantees on the input.. `json.stringify` is a great idea, however there are some more edge cases. For example, if the input is `null` or empty string or even empty array, then we get zero :sob: .

I went for introducing a new function `safeMakeNumber` (see last commit) as a workaround that covers for all edge cases and at least allows us to be explicit about it. (we could also use it inside `getTimestamp`, if we go for it) 

Let me know what you think!